### PR TITLE
The Graph

### PR DESCRIPTION
**Pull request by albertov19**
This PR introduces the documentation related to The Graph and fixes some typos here and there. Main components are:

- Some new images
- `integrations/thegraph.md ` which outlines the steps to create a Subgraph with the Lottery Contract
- `node-operators/indexers/thegraph-node.md ` which outlines the steps to run a Graph Node against either a Moonbase Alpha full node or a development node
- A couple of new variables related to the lottery contract

### CODE REVIEW

File: integrations/thegraph.md [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+
+The code above can be executed as well with `yarn codegen`.
+
+### Subgraphs Core Structure
+
+In general terms, Subgraphs define the data that The Graph will index from the blockchain and the way it is stored. Subgraphs tend to have some of the following files:
+
+ - **subgraph.yaml** — is a YAML file that contains the [Subgraph's manifest](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest), that is, information related to the smart contracts being indexed by the Subgraph
+ - **schema.grapql** — is a [GraphQL schema](https://thegraph.com/docs/define-a-subgraph#the-graphql-schema) file that defines the data store for the Subgraph being created and its structure. It is written using [GraphQL interface definition schema](https://graphql.org/learn/schema/#type-language)
+ - **AssembyScript mappings** — code in TypeScript (then compiled to[AssemblyScript](https://github.com/AssemblyScript/assemblyscript)) that is used to translate event data from the contract to the entities defined in the schema
```
Comment (themacexpert) [inline integrations/thegraph.md:76]: Typo in AssemblyScript bold heading

File: integrations/thegraph.md [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+```
+
+The code above can be executed as well with `yarn codegen`.
+
+### Subgraphs Core Structure
+
+In general terms, Subgraphs define the data that The Graph will index from the blockchain and the way it is stored. Subgraphs tend to have some of the following files:
+
+ - **subgraph.yaml** — is a YAML file that contains the [Subgraph's manifest](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest), that is, information related to the smart contracts being indexed by the Subgraph
+ - **schema.grapql** — is a [GraphQL schema](https://thegraph.com/docs/define-a-subgraph#the-graphql-schema) file that defines the data store for the Subgraph being created and its structure. It is written using [GraphQL interface definition schema](https://graphql.org/learn/schema/#type-language)
```
Comment (themacexpert) [inline integrations/thegraph.md:75]: Typo in GraphQl bold heading

File: integrations/thegraph.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+The contract hosts a lottery where a player can buy a ticket for himself or gift one to another user. When 1 hour has passed, if there are 10 participants, the next player that joins the lottery will execute a function that picks the winner. All the funds stored in the contract are sent to the winner, after which a new round starts.
+
+The main functions of the contract are the following:
+
+ - **joinLottery** — no inputs. Function to enter the lottery's current round, the value (amount of tokens) sent to the contract need to be equal to the ticket price
+ - **giftTicket** —  one input: ticket's recipient address. Similar to `joinLottery` but the ticket's owner can be set to a different address
+ - **enterLottery** — one input: ticket's owner address. An internal function that handles the lottery's tickets logic. If an hour has passed and there are at least 10 participants, it calls the `pickWinner` function
+ - **pickWinner** — no inputs. An internal function that selects the lottery winner with a pseudo-random number generator (not safe, only for demostration purposes). It handles the logic of transferring funds and resetting variable for the next lottery round
```
Comment (themacexpert) [inline integrations/thegraph.md:40]: *demonstration

File: integrations/thegraph.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+ - **dataSources/source/address** — refers to the address of the contract of interest
+ - **dataSources/source/abi** — refers to where the interface of the contract is stored inside the `types` folder created with the `codegen` command
+ - **dataSources/source/startBlock** — refers to the start block from which the indexing will start. Ideally, this value should be close to the block the contract was created in. You can use [Blockscout](https://moonbase-blockscout.testnet.moonbeam.network/) to get this information by providing the contract address. For this example, the contract was created at block `{{ networks.moonbase.thegraph.block_number }}`
+ - **dataSources/mapping/file** — refers to the location of the mapping file
+ - **dataSources/mapping/entities** — refers to the definitions of the entities in the `schema.graphql` file
+ - **dataSources/abis/name** — refers to where the interface of the contract is stored inside the `types/dataSources/name`
+ - **dataSources/abis/file** — refers to the location where the `.json` file with the contratc's ABI is stored
```
Comment (themacexpert) [inline integrations/thegraph.md:129]: *contract

File: integrations/thegraph.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+ - **schema.grapql** — is a [GraphQL schema](https://thegraph.com/docs/define-a-subgraph#the-graphql-schema) file that defines the data store for the Subgraph being created and its structure. It is written using [GraphQL interface definition schema](https://graphql.org/learn/schema/#type-language)
+ - **AssembyScript mappings** — code in TypeScript (then compiled to[AssemblyScript](https://github.com/AssemblyScript/assemblyscript)) that is used to translate event data from the contract to the entities defined in the schema
+
+### Schema.graphql
+
+It is important to outline what data needs to be extracted from the events of the contract before modifying the `schema.graphql`. Schemas need to be defined considering the requirements of the dApp itself. For this example, although there is no dApp associated with the lottery, four entities are defined:
+
+ - **Round** — refers to a lottery round. It stores an index of the round, the prize awarded, the timestamp of the creating of the round, the timestamp of when the winner was drawn, and information regarding the participating tickets, which is derived from the `Ticket` entity
```
Comment (themacexpert) [inline integrations/thegraph.md:82]: *creation of the round*

**Review by themacexpert — COMMENTED** [review]
I built in on local and it looks great! Just a few typos I have pointed out.

**Review by eshaben — CHANGES_REQUESTED** [review]
When I was going through the Integrations > The Graph tutorial it kind of skipped from setting up the subgraph and understanding all of the files to deploying a subgraph. It doesn't really mention hey you have to compile this with hardhat and if testing locally have a running node. Also in the Moonlotto repo that we clone in the instructions, in the hardhat config we use "standalone" instead of "development", and the url is set to "http://127.0.0.1:9990" which could be confusing as we typically point to 9933 for development nodes.

File: integrations/thegraph.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+To use The Graph on Moonbase Alpha you have two options:
+
+ - Point your Subgraph to The Graph API
+ - Run a Graph Node against Moonbase Alpha and point your Subgraph to it. To do so, you can follow [this tutorial](/node-operators/indexers/thegraph-node/). 
+
+## The Lottery Contract
+
+For this example, a simple Lottery contract will be used. You can find the Solidity file in [this link](https://github.com/PureStake/moonlotto-subgraph/blob/main/contracts/MoonLotto.sol). 
+
+The contract hosts a lottery where a player can buy a ticket for himself or gift one to another user. When 1 hour has passed, if there are 10 participants, the next player that joins the lottery will execute a function that picks the winner. All the funds stored in the contract are sent to the winner, after which a new round starts.
```
[the remark this answers is not in the document]
Comment (albertov19) [reply integrations/thegraph.md:33]: I changed it to "where players can buy tickets for themselves..."

File: integrations/thegraph.md (hunk 1 above)
Comment (mario-purestake) [inline integrations/thegraph.md]: Should a space be between `to[AssemblyScript]`?
```suggestion
 - **AssembyScript mappings** — code in TypeScript (then compiled to [AssemblyScript](https://github.com/AssemblyScript/assemblyscript)) that is used to translate event data from the contract to the entities defined in the schema
```

File: integrations/thegraph.md [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+
+### Subgraphs Core Structure
+
+In general terms, Subgraphs define the data that The Graph will index from the blockchain and the way it is stored. Subgraphs tend to have some of the following files:
+
+ - **subgraph.yaml** — is a YAML file that contains the [Subgraph's manifest](https://thegraph.com/docs/define-a-subgraph#the-subgraph-manifest), that is, information related to the smart contracts being indexed by the Subgraph
+ - **schema.grapql** — is a [GraphQL schema](https://thegraph.com/docs/define-a-subgraph#the-graphql-schema) file that defines the data store for the Subgraph being created and its structure. It is written using [GraphQL interface definition schema](https://graphql.org/learn/schema/#type-language)
+ - **AssembyScript mappings** — code in TypeScript (then compiled to[AssemblyScript](https://github.com/AssemblyScript/assemblyscript)) that is used to translate event data from the contract to the entities defined in the schema
+
+### Schema.graphql
```
Comment (mario-purestake) [inline integrations/thegraph.md:84]: Title as `GraphQL Schema`
```suggestion
### GraphQL Schema
```

File: integrations/thegraph.md [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+  let player = Player.load(playerId);
+  // if player doesn't exists, create it
+  if (player == null) {
+    player = new Player(playerId);
+  }
+  player.address = event.params.player;
+
+  player.save();
+
+  // ID for the ticket (round - player_address - ticke_index_round):
```
Comment (mario-purestake) [inline integrations/thegraph.md:214]:
```suggestion
  // ID for the ticket (round - player_address - ticket_index_round):
```

File: integrations/thegraph.md [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+}
+```
+
+## Deploying a Subgraph
+
+If you are going to use The Graph API (hosted service), you need to (all steps can be found in [this link](https://thegraph.com/docs/deploy-a-subgraph)):
+
+ - Create a [Graph Explorer](https://thegraph.com/explorer/) account, you will need a Github account
+ - Go to your dashboard and write down the access token
+ - If using The Graph API, you need to create your Subgraph via the "Add Subgraph" button in the Graph Explorer site
```
Comment (mario-purestake) [inline integrations/thegraph.md:235]: You refer twice here to use The Graph API when it's already its section.
```suggestion
 - Create your Subgraph via the "Add Subgraph" button in the Graph Explorer site
```

File: integrations/thegraph.md [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+
+```
+npx graph create <username>/<subgraphName> --node <graph-node>
+```
+
+Where:
+
+ - **username** — refers to the username related to the Subgraph being created
+ - **subraphName** — refers to the Subgraph name
+ - **graph-node** — refers to the URL of the hosted service to use. Typically, for a local Graph Node is `http://127.0.0.1:8020
```
Comment (mario-purestake) [inline integrations/thegraph.md:250]: Missing trailing quote
```suggestion
 - **graph-node** — refers to the URL of the hosted service to use. Typically, for a local Graph Node is `http://127.0.0.1:8020`
```

File: integrations/thegraph.md [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+Once created, you can deploy your Subgraph by running the following command with the same parameters as before:
+
+```
+npx graph deploy <username>/<subgraphName> --ipfs <ipfs-url> --node <graph-node> --access-token <access-token>
+```
+
+Where:
+
+ - **username** — refers to the username used when creating the Subgraph
+ - **subraphName** — refers to the Subgraph name defined when creating the Subgraph
```
Comment (mario-purestake) [inline integrations/thegraph.md:261]: Typo on `subgraphName`
```suggestion
 - **subgraphName** — refers to the Subgraph name defined when creating the Subgraph
```

File: node-operators/indexers/thegraph-node.md [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+description: Build APIs using The Graph indexing protocol on Moonbeam
+---
+
+# Running a Graph Node on Moonbeam
+
+![The Graph Node on Moonbeam](/images/thegraph/thegraphnode-banner.png)
+
+## Introduction
+
+A Graph Node sources events from a blockchain to deterministically update a data store that can be queried via the GraphQL endpoint.
```
Comment (mario-purestake) [inline node-operators/indexers/thegraph-node.md:12]:
```suggestion
A Graph Node sources events from a blockchain to deterministically update a data store that can be then queried via a GraphQL endpoint.
```

File: node-operators/indexers/thegraph-node.md [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    volumes:
+      - ./data/ipfs:/data/ipfs
+  postgres:
+    image: postgres
+    ports:
+      - '5432:5432'
+    command: ["postgres", "-cshared_preload_libraries=pg_stat_statements"]
+    environment:
+      POSTGRES_USER: graph-node
+      POSTGRES_PASSWORD: let-me-in
```
Comment (mario-purestake) [inline node-operators/indexers/thegraph-node.md:96]: You kept it! :laughing:

**Review by mario-purestake — CHANGES_REQUESTED** [review]

File: integrations/thegraph.md (hunk 9 above)
Comment (albertov19) [reply integrations/thegraph.md:235]: good catch

File: node-operators/indexers/thegraph-node.md (hunk 13 above)
Comment (albertov19) [reply node-operators/indexers/thegraph-node.md:96]: Why not? haha

**Comment by albertov19** [comment]
I've pushed a commit fixing the typos and I've rearranged the writing (and the moonlotto repo) so that now people can follow the steps without needing to do anything  on the Hardhat side.

File: node-operators/indexers/thegraph-node.md [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+![The Graph Node on Moonbeam](/images/thegraph/thegraphnode-banner.png)
+
+## Introduction
+
+A Graph Node sources events from a blockchain to deterministically update a data store that can be then queried via a GraphQL endpoint.
+
+There are two ways you can set up a Graph Node: you can use Docker to run an all-in-one image, or you can run their [Rust implementation](https://github.com/graphprotocol/graph-node). The steps described in this guide will only cover the Docker alternative, as it is more convenient, and you can set up a Graph Node very quickly.
+
+!!! note
+    The steps described in this guide assume an Ubuntu 18.04-based environment, and they will need to be adapted accordingly for other systems.
```
Comment (eshaben) [inline node-operators/indexers/thegraph-node.md:17]: oh just noticed this... I ran through these steps on MacOS and it works! So you can use the text snippet for this

**Review by eshaben — APPROVED** [review]
LGTM ✨  just made one last comment

**Review by mario-purestake — APPROVED** [review]
LGTM :grin:

**Pull request merged**